src/libraries/System.Runtime/tests/System/GCTests.cs

```diff
@@ -1058,11 +1058,52 @@ private static void AllocateArrayTooLarge()
             Assert.Throws<OutOfMemoryException>(() => GC.AllocateUninitializedArray<double>(int.MaxValue, pinned: true));
         }
 
-        [Fact]
-        private static void AllocateArrayRefType()
+        struct EmbeddedValueType<T>
+        {
+            unsafe fixed byte _[7];
+            public T Value;
+        }
+
+        [Theory]
+        [InlineData(false), InlineData(true)]
+        private static void AllocateArray_UninitializedOrNot_WithManagedType_DoesNotThrow(bool pinned)
+        {
+            void TryType<T>()
+            {
+                GC.AllocateUninitializedArray<T>(100, pinned);
+                GC.AllocateArray<T>(100, pinned);
+
+                GC.AllocateArray<EmbeddedValueType<T>>(100, pinned);
+                GC.AllocateUninitializedArray<EmbeddedValueType<T>>(100, pinned);
+            }
+
+            TryType<string>();
+            TryType<object>();
+        }
+
+        [Theory]
+        [InlineData(false), InlineData(true)]
+        private unsafe static void AllocateArrayPinned_ManagedValueType_CanRoundtripThroughPointer(bool uninitialized)
         {
-            GC.AllocateUninitializedArray<string>(100);
-            Assert.Throws<ArgumentException>(() => GC.AllocateUninitializedArray<string>(100, pinned: true));
+            const int length = 100;
+            var rng = new Random(0xAF);
+
+            var array = uninitialized ? GC.AllocateUninitializedArray<EmbeddedValueType<string>>(length, pinned: true) : GC.AllocateArray<EmbeddedValueType<string>>(length, pinned: true);
+            byte* pointer = (byte*)Unsafe.AsPointer(ref array[0]);
+            var size = Unsafe.SizeOf<EmbeddedValueType<string>>();
+
+            for(int i = 0; i < length; ++i)
+            {
+                var idx = rng.Next(length);
+                ref var evt = ref Unsafe.AsRef<EmbeddedValueType<string>>(pointer + size * idx);
+
+                var stringValue = rng.NextSingle().ToString();
```

Comment:
```suggestion
            int size = Unsafe.SizeOf<EmbeddedValueType<string>>();

            for(int i = 0; i < length; ++i)
            {
                int idx = rng.Next(length);
                ref EmbeddedValueType<string> evt = ref Unsafe.AsRef<EmbeddedValueType<string>>(pointer + size * idx);

                string stringValue = rng.NextSingle().ToString();
```
Nit: Repo coding conventions. (https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md - We only use var when the type is explicitly named on the right-hand side.)